doc/design/persistent_volumes_etcd_data.md

```diff
@@ -0,0 +1,125 @@
+
+User may desire consistency over availability and prefer to wait for the cluster the return in a quorate state instead of restoring from a backup (that will contain old data).
+
+This proposal aims to fix cases C and D permitting to the cluster the return in a quorate state without the need to restore from a backup.
+
+## Proposed Changes
+
+#### Changes
+
+The change to be effective and cover the above failure scenario is split in two parts.
+
+### Part 1
+
+- Add to the cluster spec PodPolicy an option to enable putting etcd data inside a Persistent Volume
+- If podPolicy has persistent volumes enabled, when creating a new member also create a Persistent Volume Claim and use it as the volume source for etcd-data
+- Handle etcd data persistent volume claims cleanup:
+  - set on the PVC an owner like already done with other resources.
+  - in the reconcile loop, remove PVCs that don't have a matching etcd member in the etcd cluster and their related pod doesn't exists in the k8s API
+  - garbage collect also PVCs
+
+#### Pod name to Persistent Volume Claim name mapping
+
+The pod and related PVC are mapped 1 to 1 so their names be easily generated and deducted. I.E. if the pod is called `etcd-0000` its related PVC name will be `etcd-0000-pvc`
+
+#### Results
+
+These changes will put the basis for initial persistent volume management
+
+#### Problems
+
+This first part won't fix case C and D.
+
+### Part 2
+
+#### Changes
+
+- Change reconcile logic to handle deleted pods that had a persistent volume:
+
+If a pod of an etcd cluster member is deleted and it had a persistent volume claim associated, instead of removing the etcd member from the etcd cluster and adding a new member just recreate it (keeping its pod name) binding the existing persistent volume claim. The pod recreation will happen only if the pod doesn't exists anymore in the API because we'll keep the same pod name. This also enable a MANDATORY property: at most once pod existence logic to avoid having in the API at the same time two pods with associated the same PVC.
+
+#### Results
+
+With this change if a majority of pods are deleted the etcd-operator will recreate them keeping their previous data, so case C and D will be fixed.
+
+
+#### Problems
+
+- until the pod isn't removed from the api server it won't be recreated (for some more details see the [Pod Safety, Consistency Guarantees, and Storage Implications](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/pod-safety.md) k8s design doc)
+- if the etcd data gets corrupted etcd will exit and the pod will go in a failed state and then removed from the api server. The change introdudec in this part will make etcd-operator recreate the pod with the previous corrupted data in a loop.
+- there can be other unpredictable cases where etcd will continuously fail to start like in the above point (fs full, os problems etc...).
```

Comment:
This should be covered as pod failure.

Reply:
We also need to cover node failure in part 2.

Reply:
> We also need to cover node failure in part 2.

Are you talking about k8s nodes?

Reply:
Yeah. I categorize them in https://github.com/coreos/etcd-operator/pull/1374#discussion_r137345989 .